jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java

```diff
@@ -226,113 +242,93 @@ protected void stop(LifeCycle l) throws Exception
     @Override
     protected void doStop() throws Exception
     {
-        try (AutoLock l = lock())
+        try (AutoLock ignored = lock())
         {
             super.doStop();
 
             // Stop filters
             List<FilterHolder> filterHolders = new ArrayList<>();
-            List<FilterMapping> filterMappings = ArrayUtil.asMutableList(_filterMappings);
-            if (_filters != null)
+            for (int i = _filters.size(); i-- > 0; )
             {
-                for (int i = _filters.length; i-- > 0; )
+                FilterHolder filter = _filters.get(i);
+                try
                 {
-                    FilterHolder filter = _filters[i];
-                    try
-                    {
-                        filter.stop();
-                    }
-                    catch (Exception e)
-                    {
-                        LOG.warn("Unable to stop filter {}", filter, e);
-                    }
-                    if (filter.getSource() != Source.EMBEDDED)
-                    {
-                        //remove all of the mappings that were for non-embedded filters
-                        _filterNameMap.remove(filter.getName());
-                        //remove any mappings associated with this filter
-                        filterMappings.removeIf(fm -> fm.getFilterName().equals(filter.getName()));
-                    }
-                    else
-                        filterHolders.add(filter); //only retain embedded
+                    filter.stop();
+                }
+                catch (Exception e)
+                {
+                    LOG.warn("Unable to stop filter {}", filter, e);
+                }
+                if (_durable.contains(filter))
+                {
+                    filterHolders.add(filter); //only retain durable
                 }
             }
 
-            //Retain only filters and mappings that were added using jetty api (ie Source.EMBEDDED)
-            FilterHolder[] fhs = filterHolders.toArray(new FilterHolder[0]);
-            updateBeans(_filters, fhs);
-            _filters = fhs;
-            FilterMapping[] fms = filterMappings.toArray(new FilterMapping[0]);
-            updateBeans(_filterMappings, fms);
-            _filterMappings = fms;
+            //Retain only durable filters and mappings
+            updateBeans(_filters, filterHolders);
+            _filters.clear();
+            _filters.addAll(filterHolders);
 
-            _matchAfterIndex = (_filterMappings.length == 0 ? -1 : _filterMappings.length - 1);
+            // Update indexes for prepending filters
```

Comment:
I don't see any code here that removes the non-durable filtermappings, yet we've set the matchBefore and matchAfter index.